src/Http/Http.Extensions/src/RequestDelegateFactory.cs

```diff
@@ -292,6 +297,60 @@ private static RouteHandlerFilterDelegate CreateFilterPipeline(MethodInfo method
         return filteredInvocation;
     }
 
+    private static Expression MapHandlerReturnTypeToValueTask(Expression methodCall, Type returnType)
+    {
+        if (returnType == typeof(void) || returnType == typeof(Task) || returnType == typeof(ValueTask))
+        {
+            return Expression.Block(methodCall, Expression.Constant(new ValueTask<object?>(EmptyHttpResult.Instance)));
```

Comment:
This isn't waiting for the delegate to complete before returning. You can see the behavior by having the following code:
```csharp
app.MapGet("/task", handler: async (HttpContext c) =>
{
    await Task.Delay(10000);
    c.Response.StatusCode = 400;
})
.AddFilter((c, n) =>
{
    return n(c);
});
```

A request to `/task` will complete immediately with a 200, and after 10 seconds VS will show an ODE in the Debug output window (this will be an unobserved exception since no is waiting for the completion anymore).

Then, if you remove the filter you will see a request to `/task` will complete after 10 seconds and have a status of 400 as expected.

Reply:
Good catch! I just pushed c1cca8b91d13d28e9ca94bbd57d690378e86c5e3 with a fix for this.